### PR TITLE
add preferred client mode

### PR DESCRIPTION
**Pull request by lukasmatusiewicz**
- make use of `preferred_client_mode`
- update tests
- fix dependencies
- update naming as provided by the convention
- handle lost `multi_challenge` data
- convert `ServiceAccountAvailable` method to the inline code (only one usage)
- boundle GetAuthToken's content into an if-else statement

Closing #3

### CODE REVIEW

**Comment by lukasmatusiewicz** [comment]
@nilsbehlen , should this commented code be removed, or fixed, or should stay as it is?
https://github.com/privacyidea/csharp-client/blob/main/PrivacyIDEA_Client/PrivacyIDEA.cs#L327

File: PrivacyIDEA_Client/PIResponse.cs [hunk 1]
```diff
@@ -150,41 +154,47 @@ public List<string> WebAuthnSignRequests()
...
+                                        TransactionID = transactionid,
+                                        Type = type
+                                    };
+                                    ret.Challenges.Add(tmp);
+                                }
                             }
                             else
                             {
-                                PIChallenge tmp = new()
+                                if (privacyIDEA != null)
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PIResponse.cs:194]: use pattern matching here too, should be consistent everywhere.

File: PrivacyIDEA_Client/PIResponse.cs [hunk 2]
```diff
@@ -150,41 +154,47 @@ public List<string> WebAuthnSignRequests()
...
-                                    var signRequest = attr["webAuthnSignRequest"];
-                                    if (signRequest != null)
+                                    PIWebAuthnSignRequest tmp = new();
+
+                                    if (element["attributes"] is JToken attr && attr.Type != JTokenType.Null)
                                     {
-                                        tmp.WebAuthnSignRequest = signRequest.ToString(Formatting.None);
-                                        tmp.WebAuthnSignRequest.Replace("\n", "");
+                                        var signRequest = attr["webAuthnSignRequest"];
+                                        if (signRequest != null)
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PIResponse.cs:167]: use pattern matching

File: PrivacyIDEA_Client/PIResponse.cs [hunk 3]
```diff
@@ -129,7 +129,11 @@ public List<string> WebAuthnSignRequests()
 
                 if (jobj["detail"] is JToken detail && detail.Type != JTokenType.Null)
                 {
-                    if (detail["transaction_id"] is not null) // todo if null ignore or error?
+                    if (detail["preferred_client_mode"] is not null)
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PIResponse.cs:132]: if (detail["preferred_client_mode"] is string pref_client_mode)

File: PrivacyIDEA_Client/PIResponse.cs [hunk 4]
```diff
@@ -129,7 +129,11 @@ public List<string> WebAuthnSignRequests()
...
+                    {
+                        ret.PreferredClientMode = (string)detail["preferred_client_mode"]!;
+                    }
+                    if (detail["transaction_id"] is not null)
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PIResponse.cs:136]: same as above

File: PrivacyIDEA_Client/PrivacyIDEA.cs [hunk 5]
```diff
@@ -297,69 +298,73 @@ public PIResponse ValidateCheck(string user, string otp, string? transactionid =
...
+                // todo rm?
+
+                JObject root = JObject.Parse(response);
+                if (root["result"] is JToken result)
                 {
-                    if (tkn["token"] is not null)
+                    if (result["value"] is JToken tkn)
                     {
-                        token = (string)tkn["token"]!; // todo how the response looks like? should'nt it be one floor above?
+                        if (tkn["token"] is not null)
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PrivacyIDEA.cs:345]: pattern matching

File: PrivacyIDEA_Client/PrivacyIDEA.cs [hunk 6]
```diff
@@ -297,69 +298,73 @@ public PIResponse ValidateCheck(string user, string otp, string? transactionid =
...
+                /*      try
+                      {
+                          dynamic root = JsonConvert.DeserializeObject(response);
+                          token = root.result.value.token;
+                      }
+                      catch (Exception)
+                      {
+                          Error("/auth response did not have the correct format or did not contain a token.\n" + response);
+                      }*/
+                // todo rm?
```
Comment (nilsbehlen) [inline PrivacyIDEA_Client/PrivacyIDEA.cs:338]: you can remove the old code

File: Tests/Tests.csproj [hunk 7]
```diff
@@ -1,23 +1,35 @@
...
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="WireMock.Net" Version="1.5.10" />
   </ItemGroup>
 
 	<ItemGroup>
 		<ProjectReference Include="..\PrivacyIDEA_Client\PrivacyIDEA_Client.csproj" />
 	</ItemGroup>
 
+	<ItemGroup>
```
Comment (nilsbehlen) [inline Tests/Tests.csproj:29]: i don't think this file is needed - if it is, there is a problem somehwere

**Review by nilsbehlen — CHANGES_REQUESTED** [review]
remove the Microsoft.IdentityServer.Web.dll please

**Review by nilsbehlen — APPROVED** [review]

**Pull request merged**